docs/manpages/ntvquerygen.1
.\" Automatically generated by Pod::Man version 1.15
.\" Fri Nov 22 12:11:10 2002
.\"
.\" Standard preamble:
.\" ======================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Ip \" List item
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R

.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used
.\" to do unbreakable dashes and therefore won't be available.  \*(C` and
.\" \*(C' expand to `' in nroff, nothing in troff, for use with C<>
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr
.\" for titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and
.\" index entries marked with X<> in POD.  Of course, you'll have to process
.\" the output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it
.\" makes way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.bd B 3
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ======================================================================
.\"
.IX Title "NTVQUERYGEN 1"
.TH NTVQUERYGEN 1 "2.0.0" "2002-11-22" "NexTrieve"
.UC
.SH "NAME"
ntvquerygen \- Generate query \s-1XML\s0 from parameters
.SH "SYNPOSIS"
.IX Header "SYNPOSIS"
.Vb 9
\& ntvquerygen -[?xX]
\&              [-i offset] [-d displayedhits] [-t totalhits]
\&              [-v n] [-e {utf8|utf16|iso|ascii}]
\&              [-f fuzziness] [-h highlight-length]
\&              [-T texttype] [-sa[=0|1]] [-sp[=0|1]]
\&              [-fww=n] [-fwv=n] [-D idxname]
\&              [-c constraint]
\&              [-any="text"] [-all="text"] [-not="text"]
\&              [more query text...]
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBntvquerygen\fR allows the generationg of nextrieve query \s-1XML\s0 from
parameters.  The query \s-1XML\s0 is printed to standard output.
.PP
A definition of what the \s-1XML\s0 for a query looks like
can be found in \fIntvqueryxml\fR\|(5).
.PP
An example usage (reformatted) is:
.PP
.Vb 11
\&    $ ntvquerygen hello
\&    <?xml version="1.0" encoding="ISO-8859-1"?>
\&    <ntv:query xmlns:ntv="http://www.nextrieve.com/1.0">hello</ntv:query>
\&    $ ntvquerygen -x -T subjects -c financial blackbird
\&    <?xml version="1.0" encoding="ISO-8859-1"?>
\&    <ntv:query xmlns:ntv="http://www.nextrieve.com/1.0" type="exact">
\&        <constraint>financial</constraint>
\&        <texttype name="subjects" weight="100"/>
\&        blackbird
\&    </ntv:query>
\&    $
.Ve
.SH "OPTIONS"
.IX Header "OPTIONS"
.Ip "\fB\-?\fR" 4
.IX Item "-?"
Write extended help information to standard output and exit.
.Ip "\fB\-v 0|1\fR" 4
.IX Item "-v 0|1"
Sets the 'longform' query attribute, to have a long form of the
hitlist returned.
.Ip "\fB\-x\fR" 4
.IX Item "-x"
Set the 'type=\*(L"exact\*(R"' query attribute, indicating an exact search.
.Ip "\fB\-i offset\fR" 4
.IX Item "-i offset"
Set the 'firsthit="\fIoffset\fR"' query attribute, indicating the 1\-based
offset from where to return hits in the hitlist.
.Ip "\fB\-d displayed-hits\fR" 4
.IX Item "-d displayed-hits"
Set the 'displayedhits="\fIdisplayed-hits\fR"' query attribute, indicating how
many hits to return.
.Ip "\fB\-t total-hits\fR" 4
.IX Item "-t total-hits"
Set the 'totalhits="\fItotal-hits\fR"' query attribute, indicating the maximum
number of hits to process.
.Ip "\fB\-e {utf8|utf16|iso|ascii}\fR'" 4
.IX Item "-e {utf8|utf16|iso|ascii}'"
The encoding is indicated.  An initial line like
.Sp
.Vb 1
\&    <?xml version="1.0" encoding="ISO-8859-1"?>
.Ve
is printed.  Note that this encoding is only applied to query \s-1XML\s0
\&\-\- the text present in the text index is always encoded with \s-1UTF-8\s0.
.Ip "\fB\-f fuzziness\fR" 4
.IX Item "-f fuzziness"
Set the 'fuzzylevel="\fIfuzziness\fR"' query attribute (in the range 0 to 3
inclusive), indicating the level of fuzziness for a fuzzy search.
.Ip "\fB\-h highlight-length\fR" 4
.IX Item "-h highlight-length"
Set the 'highlightlength="\fIhighlight-length\fR"' query attribute,
indicating the minimum length a word must have in the preview before
it's highlighted if it's relevant to the fuzzy search query performed.
.Ip "\fB\-T \f(BItexttype[=weightpercentage]\fB\fR" 4
.IX Item "-T texttype[=weightpercentage]"
Set a <texttype> element in the query container, indicating a texttype
specification.  Special values for \fItexttype\fR are empty ("\*(L")
indicating the default text type, and \*(R"*\*(L" indicating \*(R"all other text types".
.Ip "\fB\-sa[=0|1]\fR" 4
.IX Item "-sa[=0|1]"
Set the 'showattributes="\fIn\fR"' query attribute, indicating whether
attribute information should be returned with hits or not.
.Ip "\fB\-sp[=0|1]\fR" 4
.IX Item "-sp[=0|1]"
Set the 'showpreview="\fIn\fR"' query attribute, indicating whether
preview information should be returned with hits or not.
.Ip "\fB\-D idxname\fR" 4
.IX Item "-D idxname"
Set the <indexname> element in the query container, indicating what
logical index to search if we're communicating with a nextrieve
caching server.
.Ip "\fB\-c constraint\fR" 4
.IX Item "-c constraint"
Set the <constraint> element in the query container, indicating a
constraint to apply to the search.
.Ip "\fB\-any=text\fR, \fB\-all=text\fR, \fB\-not=text\fR, \fBmore query text\fR" 4
.IX Item "-any=text, -all=text, -not=text, more query text"
Set one of the <qany>, <qall>, or <qnot> elements in the query container,
or place \fImore query text\fR directly in the query container.
.Sp
These represent classes of text to search for.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
.Vb 1
\&    ntvsearch(1), ntvsearchd(1), ntvresources(5), ntvcached(1).
.Ve


src/ntvquerygen.c
/*
Copyright (c) 2003 Kim Hendrikse

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

#include <config.h>

#ifdef WIN32
#include "StdAfx.h"
#endif

#include <stdio.h>
#include <limits.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include <ctype.h>
#include <stdarg.h>

#if defined(USING_THREADS)
#include <pthread.h>
#if defined(USING_SEMAPHORE_H)
#include <semaphore.h>
#else
#include "bsdsem.h"
#endif
#endif

#include "ntvstandard.h"
#include "ntvrrw.h"
#include "rbt.h"
#include "ntvreq.h"
#include "ntvquery.h"
#include "ntvmemlib.h"
#include "ntvutils.h"

#include "getopt.h"

reqbuffer_t req_default; /* Filled with input defaults. */

/*
 * Print usage message
 */
void usage( int c, char *filename )
{
    printf("Usage: %s", filename);

    printf
	(
	    " -[?xX]"
	    " [-c constraint] [-r rankingattribute]\n"
	    " [-i offset] [-d displayed-hits] [-t total-hits]\n"
	    " [-v n] [-e {utf8|utf16|iso|ascii}]"
	    " [-f fuzziness]"
	    " [-h highlight-length]\n"
	    " [-T texttype]"
	    " [-sa[=0|1]] [-sp[=0|1]] [-tr[=0|1]]\n"
	    " [-fww=n] [-fwv=n]"
	    " [-D dbname]\n"
	    " [-any=\"text\"] [-all=\"text\"] [-not=\"text\"] [query text...]\n"
	);

    if ( c == '?' )
    {
	printf
	(
	"\n"
	"-?: print this list.\n"
	"-any=\"words\": search for some words.\n"
	"-all=\"words\": all of these words must be present in returned documents.\n"
	"-not=\"words\": none of these words must be present in returned documents.\n"
	"-fww=fuzzy-word-weight-percent: default 20.\n"
	"-fwv=fuzzy-word-variation: defaults from resource file.\n"
	"-D database-discriminator: name of logical db to search.\n"
	"-x: perform exact search.\n"
	"-X: perform exact search using document-level information only.\n"
	"-sa[=0|1]: show attribute info in returned hit list.\n"
	"-sp[=0|1]: show previews in returned hit list.\n"
	"-tr[=0|1]: use text for rating.\n"
	"-enc=utf8|utf16|iso|ascii: specify query text encoding.\n"
	"-id=id: give query an id that is returned with hit list.\n"
	"-c constraint: specify constraint string.\n"
	"-r ranking: specify (floating point) attribute for ranking.\n"
	"-v #: long hitlist tag names (-v 1) or short (-v 0).\n"
	"-i #: index (starting from 1) of first hit to display.\n"
	"-d #: number of hits to display.\n"
	"-t #: total number of hits to rate.\n"
	"-f #: fuzzy factor.\n"
	"-h #: minimum word length to highlight in previews.\n"
	"-T texttype[=weight-percent]: restrict search to text types.  Can be\n"
        "   repeated.  texttype of '*' implies all non-specified types.\n"
	"\n"
	"query: search query.\n"
	"       +word: word is required to be in returned documents.\n"
	"       -word: word is required to NOT be in returned documents.\n"
	"\n"
	);
    }

    if ( c == '?' )
    	exit( 0 );
    else
    	exit( 1 );
}


/*
 * Ensure is a digit
 */
static void digitCheck( char message[], char input[] )
{
    char *s;

    if ( !*input ) {
	fprintf( stderr, "%s", message );
	exit( 1 );
    }

    for ( s = input; *s; s++ )
	if ( !isdigit( *s&0xff ) ) {
	    fprintf( stderr, "%s", message );
	    exit( 1 );
	}
}


int main( int argc, char **argv )
{
    unsigned long ch;
    extern int optind;
    extern char *optarg;

    unsigned char *ttstuff;
    unsigned char *wstuff;
    int w;

#define Q_ANY 1000
#define Q_ALL 1001
#define Q_NOT 1002
#define Q_FWW 1003
#define Q_FLV 1004
#define Q_ID  1005
#define Q_SA  1006
#define Q_SP  1007
#define Q_TR  1008
#define Q_ENC 1009

    struct option opts[] =
	{
	    {"any", required_argument, NULL, Q_ANY},
	    {"all", required_argument, NULL, Q_ALL},
	    {"not", required_argument, NULL, Q_NOT},
	    {"fww", required_argument, NULL, Q_FWW},
	    {"fwv", required_argument, NULL, Q_FLV},
	    {"id",  required_argument, NULL, Q_ID},
	    {"sa",  optional_argument, NULL, Q_SA},
	    {"sp",  optional_argument, NULL, Q_SP},
	    {"tr",  optional_argument, NULL, Q_TR},
	    {"enc", required_argument, NULL, Q_ENC},
	    {"help", no_argument, NULL, '?'},
	    {NULL, no_argument, NULL, 0}
	};

    /* Set environment variables */
    memset(&req_default, 0, sizeof(req_default));
    req_init_hdrinfo(&req_default, NULL);
    req_default.encoding = "ISO-8859-1";

    while ( ( ch = getopt_long_only( argc, argv, "?qXxCVv:L:I:R:A:T:N:c:r:P:l:f:t:h:d:u:i:F:D:", opts, NULL ) ) != EOF)
	switch ( ch )
	{
	case Q_ANY:
	    if (req_default.qryAnyStrLen > 0)
		ntvStrAppend
		    (
			" ", 1,
			&req_default.qryAnyStr,
			&req_default.qryAnyStrSz, &req_default.qryAnyStrLen
		    );
	    ntvStrAppend
		(
		    optarg, -1,
		    &req_default.qryAnyStr,
		    &req_default.qryAnyStrSz, &req_default.qryAnyStrLen
		);
	    break;
	case Q_ALL:
	    if (req_default.qryAllStrLen > 0)
		ntvStrAppend
		    (
			" ", 1,
			&req_default.qryAllStr,
			&req_default.qryAllStrSz, &req_default.qryAllStrLen
		    );
	    ntvStrAppend
		(
		    optarg, -1,
		    &req_default.qryAllStr,
		    &req_default.qryAllStrSz, &req_default.qryAllStrLen
		);
	    break;
	case Q_NOT:
	    if (req_default.qryNotStrLen > 0)
		ntvStrAppend
		    (
			" ", 1,
			&req_default.qryNotStr,
			&req_default.qryNotStrSz, &req_default.qryNotStrLen
		    );
	    ntvStrAppend
		(
		    optarg, -1,
		    &req_default.qryNotStr,
		    &req_default.qryNotStrSz, &req_default.qryNotStrLen
		);
	    break;
	case Q_FWW:
	    req_default.ntvFuzzyWordWeight = atoi(optarg);
	    break;
	case Q_FLV:
	    req_default.ntvFuzzyLenVariation = atoi(optarg);
	    break;
	case Q_ID:
	    /*
	     * "name" -- give the query an id that's returned in the
	     * result list.
	     */
	    req_default.ntvID = STRDUP(optarg);
	    break;
	case 'v':
	    if (optarg != NULL && *optarg != 0)
	    {
		digitCheck("-v: not given a number.\n", optarg);
		req_default.ntvShowLongForm = atoi(optarg);
	    }
	    else
		req_default.ntvShowLongForm = TRUE;
	    break;
	case 'D':
	    /*
	     * Database discriminator if we're talking to a caching
	     * server.
	     */
	    req_default.ntvDBName = STRDUP(optarg);
	    req_default.ntvDBNameLen = strlen(optarg);
	    req_default.ntvDBNameSz = req_default.ntvDBNameLen+1;
	    break;
	case 'x' :
	    req_default.ntvSearchType = NTV_SEARCH_EXACT;
	    break;
	case 'X':
	    req_default.ntvSearchType = NTV_SEARCH_DOCLEVEL;
	    break;
	case Q_SA:
	    if (optarg == NULL || *optarg == 0)
		req_default.ntvShowAttributes = TRUE;
	    else
		req_default.ntvShowAttributes = !!atoi(optarg);
	    break;
	case Q_SP:
	    if (optarg == NULL || *optarg == 0)
		req_default.ntvShowPreviews = TRUE;
	    else
		req_default.ntvShowPreviews = !!atoi(optarg);
	    break;
	case Q_TR:
	    if (optarg == NULL || *optarg == 0)
		req_default.ntvTextRate = TRUE;
	    else
		req_default.ntvTextRate = !!atoi(optarg);
	    break;
	case Q_ENC:
	    /* encoding. */
	    if (strcmp(optarg, "utf8") == 0)
		req_default.encoding = "UTF-8";
	    else if (strcmp(optarg, "utf16") == 0)
		req_default.encoding = "UTF-16";
	    else if (strcmp(optarg, "iso") == 0)
		req_default.encoding = "ISO-8859-1";
	    else if (strcmp(optarg, "ascii") == 0)
		req_default.encoding = "US-ASCII";
	    else
	    {
		fprintf
		    (
			stderr,
			"invalid encoding: \"%s\"\n"
			"   use one of: utf8, utf16, iso, ascii.\n",
			optarg
		    );
		exit(1);
	    }
	    break;
	case 'c' :
	    req_default.constraintString = STRDUP(optarg);
	    req_default.constraintStringLen = strlen(optarg);
	    req_default.constraintStringSz = req_default.constraintStringLen+1;
	    break;
	case 'r':
	    req_default.rankingString = STRDUP(optarg);
	    req_default.rankingStringLen = strlen(optarg);
	    req_default.rankingStringSz = req_default.rankingStringLen+1;
	    break;
	case 'i' :
	    digitCheck( "-i: Hit offset must be a number\n", optarg );
	    req_default.ntvOffset = atol(optarg);
	    break;
	case 'f' :
	    digitCheck("-f: Fuzzy factor must be a number\n", optarg);
	    req_default.ntvFuzzyFactor = atol( optarg );
	    break;
	case 't' :
	    lowerit( trim( shiftleft( optarg ) ) );
	    if (strcmp( optarg, "all") == 0)
		req_default.ntvTotalScores = ULONG_MAX;
	    else
	    {
		digitCheck("-t: Total hits must be a number or \"all\"\n", optarg);
		req_default.ntvTotalScores = atol(optarg);
	    }
	    break;
	case 'h' :
	    digitCheck("-h: Highlighting must be a number\n", optarg);
	    req_default.ntvHighlightChars = atol( optarg );
	    break;
	case 'd' :
	    lowerit( trim( shiftleft( optarg ) ) );
	    if ( !strcmp( optarg, "all" ) )
		req_default.ntvDisplayedHits = ULONG_MAX;
	    else
	    {
		digitCheck("-d: Displayed hits must be a number or \"all\"\n", optarg);
		req_default.ntvDisplayedHits = atol( optarg );
	    }
	    break;
	case 'T':
	    /* Text type spec. */
	    ttstuff = optarg;
	    if((wstuff = strchr(ttstuff, '=')) != NULL)
	    {
		w = atoi(wstuff+1);
		*wstuff = 0; /* terminate text type name. */
	    }
	    else
		w = -1;
	    /* Space trim arguments for niceness. */
	    while (isspace(*ttstuff))
		ttstuff++;
	    wstuff = ttstuff+strlen(ttstuff)-1;
	    while (wstuff >= ttstuff && isspace(*wstuff))
		*wstuff-- = 0;
	    if (!req_addtexttypespec(&req_default, ttstuff, w))
		exit(1);
	    break;
	case '?' :
	    usage( optopt, *argv );
	default :
	    usage( 0, *argv );
	}

    argc -= optind;

    if
	(
	    argc <= 0
	    && req_default.constraintString == NULL
	    && req_default.qryAllStrLen == 0
	    && req_default.qryAnyStrLen == 0
	)
    {
	usage( 0, *argv );
    }

    argv += optind;
	
    while ( argc > 0 )
    {
	if (req_default.qryFrfStrLen > 0)
	    ntvStrAppend
		(
		    " ", -1,
		    &req_default.qryFrfStr,
		    &req_default.qryFrfStrSz, &req_default.qryFrfStrLen
		);
	ntvStrAppend
	    (
		*argv, -1,
		&req_default.qryFrfStr,
		&req_default.qryFrfStrSz, &req_default.qryFrfStrLen
	    );
	argc--;  argv++;
    }

    req_default.rrw = rrwClientFileIO(NULL, stdout);

    /* Simply generates XML and writes to stdout now. */
    ntvClientWriteQuery(&req_default);
    exit(0);
}
